source/Octopus.Cli/Util/RepositoryExtensions.cs

```diff
@@ -0,0 +1,145 @@
+        {
+            TResource resourceById;
+            if (!Regex.IsMatch(nameOrId, $@"^{Regex.Escape(fixedIdPrefix)}-\d+$",
+                RegexOptions.Compiled | RegexOptions.IgnoreCase))
+            {
+                resourceById = null;
+            }
+            else
+            {
+                try
+                {
+                    resourceById = await repository.Get(nameOrId)
+                        .ConfigureAwait(false);
+                }
+                catch (OctopusResourceNotFoundException)
+                {
+                    resourceById = null;
+                }
+            }
+
+            var resourceByName = await findByNameFunc(nameOrId)
+                .ConfigureAwait(false);
+
+            if (resourceById == null && resourceByName == null)
+            {
+                throw new CouldNotFindException(typeDescription, nameOrId, inDescription);
+            }
+
+            if (resourceById != null
+                && resourceByName != null
+                && !string.Equals(resourceById.Id, resourceByName.Id, StringComparison.OrdinalIgnoreCase))
+            {
+                throw new CommandException(
+                    $"Ambiguous {typeDescription} reference '{nameOrId}' matches one {typeDescription} by name and another by id.");
+            }
+
+            var found = resourceById ?? resourceByName;
+
+            if (!skipLog)
+            {
+                Logger.Debug($"Found {typeDescription}: {found.Name} ({found.Id})");
+            }
+
+            return found;
+        }
+
+        private static async Task<TResource[]> FindByNamesOrIdsOrFail<T, TResource>(this T repository,
+            Func<string, Task<TResource>> findByNameFunc, string fixedIdPrefix, string typeDescription,
+            IEnumerable<string> namesOrIds, string inDescription = "", bool skipLog = false)
+            where T : IGet<TResource> where TResource : Resource, INamedResource
+        {
+            var results = await Task.WhenAll(namesOrIds.Select<string, Task<(string missing, TResource resource)>>(
+                async nameOrId =>
+                {
+                    try
+                    {
+                        return (null, await repository.FindByNameOrIdOrFail(findByNameFunc, fixedIdPrefix,
+                                typeDescription, nameOrId, skipLog: true, inDescription: inDescription)
+                            .ConfigureAwait(false));
+                    }
+                    catch (CouldNotFindException)
+                    {
+                        return (nameOrId, null);
+                    }
+                })).ConfigureAwait(false);
+            var missing = results.Select(r => r.missing)
+                .Where(m => m != null)
+                .ToArray();
+            if (missing.Any())
+            {
+                var missingStr = string.Join(", ", missing.Select(m => '"' + m + '"'));
+                throw new CommandException(
```

Comment:
Why are we throwing a CommandException here rather than a `CouldNotFindException`?

Reply:
Oh this. The constructor signatures made it annoying so I aborted changing it, but it's done now.